src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -498,6 +568,157 @@ ShardsMatchingDeleteCriteria(Oid relationId, List *shardIntervalList,
 
 
 /*
+ * SendQueryToShards executes the given query in all placements of the given 
+ * shard list and returns the total affected tuple count. The execution is done
+ * in a distributed transaction and the commit protocol is decided according to
+ * the value of citus.multi_shard_commit_protocol parameter. SendQueryToShards
+ * does not acquire locks for the shards so it is advised to acquire locks to
+ * the shards when necessary.
+ */
+static int
+SendQueryToShards(Query *query, List *shardIntervalList)
+{
+	ListCell *shardIntervalCell = NULL;
+	List *connectionList = NIL;
+	int affectedTupleCount = 0;
+
+	PG_TRY();
+	{
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(
+				shardIntervalCell);
+			Oid relationId = shardInterval->relationId;
+			uint64 shardId = shardInterval->shardId;
+			StringInfo shardQueryString = makeStringInfo();
+			char* shardQueryStringData = NULL;
+			
+			deparse_shard_query(query, relationId, shardId, shardQueryString);
+			shardQueryStringData = shardQueryString->data;
+			int shardAffectedTupleCount = SendQueryToPlacements(shardInterval, 
```

Comment:
It kind of feels like the TransactionConnection is the main outcome of this function and should be the return value of the function, and the shardAffectedTupleCount passed as a pointer.

Alternatively, since the list of shards is known at the start, you could add something like a connectionList = OpenRemoteTransactions(shardIdList) call for symmetry with the Prepare/Abort/Commit functions. That function could return the connection list. You could also do error handling in that function (e.g. if it fails to open one of the connections, close all of them and error out afterwards). That way, you can call this function before PG_TRY and avoid the setjmp problem.